social_engagement/handlers.py

```diff
@@ -11,107 +9,146 @@
-
-    if social_engagement_enabled:
-        if course_id and user_id:
-            task_update_user_engagement_score.delay(unicode(course_id), user_id)
+    thread = kwargs['post']
+    course_id = getattr(thread, 'course_id', None)
+    user_id = getattr(thread, 'user_id', None)
 
-        # present only in two cases: thread_deleted & comment_deleted
-        if 'involved_users' in kwargs:
-            involved_users = kwargs['involved_users']
-            users = User.objects.filter(username__in=involved_users).exclude(id=user_id)
+    # present if thread_deleted
+    if 'involved_users' in kwargs:
+        users = kwargs['involved_users']
+        for user, user_data in users.items():
+            _decrement(user, course_id, user_data)
 
-            for user in users:
-                task_update_user_engagement_score.delay(unicode(course_id), user.id)
+    # thread or comment voted
+    else:
+        change = _decrement if kwargs.get('undo') else _increment
+        change(user_id, course_id, 'num_upvotes')
 
 
 @receiver(thread_created)
 def thread_created_signal_handler(sender, **kwargs):  # pylint: disable=unused-argument
     """
-    Updates user social engagement score
+    Updates user social engagement score for created thread.
     """
     thread = kwargs['post']
     course_id = getattr(thread, 'course_id', None)
     action_user = kwargs['user']
-    social_engagement_enabled = settings.FEATURES.get('ENABLE_SOCIAL_ENGAGEMENT', False)
 
-    if social_engagement_enabled and course_id and action_user:
-        task_update_user_engagement_score.delay(unicode(course_id), action_user.id)
+    if action_user:
+        _increment(action_user.id, course_id, 'num_threads')
+
+
+@receiver(comment_deleted)
+def comment_deleted_signal_handler(sender, **kwargs):  # pylint: disable=unused-argument
+    """
+    Updates user social engagement score for deleted comment.
+    """
+    post = kwargs['post']
+    course_id = getattr(post, 'course_id', None)
+
+    if 'involved_users' in kwargs:
+        users = kwargs['involved_users']
+        for user, user_data in users.items():
+            _decrement(user, course_id, user_data)
 
 
 @receiver(comment_created)
 def comment_created_signal_handler(sender, **kwargs):  # pylint: disable=unused-argument
     """
-    Updates user social engagement score
+    Updates user social engagement score for created comment.
     """
     comment = kwargs['post']
     course_id = getattr(comment, 'course_id', None)
     thread_id = getattr(comment, 'thread_id', None)
     parent_id = getattr(comment, 'parent_id', None)
     action_user = kwargs['user']
-    social_engagement_enabled = settings.FEATURES.get('ENABLE_SOCIAL_ENGAGEMENT', False)
 
-    if social_engagement_enabled and course_id and action_user:
-        task_update_user_engagement_score.delay(unicode(course_id), action_user.id)
+    if action_user:
+        # a comment is a reply to a thread
+        # a response is a reply to a response
```

Comment:
Do you mean a response is a reply to a comment?

Reply:
That's confusing, so I guess my response (pun intended) will be to just do what seems right :p

Reply:
All right, so I'll leave `# a response is a reply to a comment or a response`, because this is how it's recursively counted during deletion.